### PR TITLE
python-common: fix /hosts/ parsing in servicespecs

### PR DESCRIPTION
**Pull request by jschmid1**
Signed-off-by: Joshua Schmid <jschmid@suse.de>

todo:
- [x] add more tests
- [x] rebase to master

### CODE REVIEW

File: src/python-common/ceph/tests/test_service_spec.py [hunk 1]
```diff
@@ -42,28 +44,50 @@ def test_parse_placement_specs(test_input, expected):
...
+      }
     }
     """
-    example = json.loads(test_rgwspec.__doc__.strip())
-    spec = RGWSpec.from_json(example)
-    assert servicespec_validate_add(spec) is None
\ No newline at end of file
+    example = yaml.load(test_servicespec_rgw.__doc__.strip())
+    spec = ServiceSpec.from_json(example)
+    assert isinstance(spec, RGWSpec)
```
Comment (sebastian-philipp) [inline src/python-common/ceph/tests/test_service_spec.py]: :+1: for merging this into a single test.

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -176,7 +178,7 @@ def __repr__(self):
     def from_json(cls, data):
         hosts = data.get('hosts', [])
         if hosts:
-            data['hosts'] = [HostPlacementSpec.from_json(host) for host in hosts]
+            data['hosts'] = [HostPlacementSpec.parse(host) for host in hosts]
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/service_spec.py]: `to_json()` also needs this change.

**Comment by jschmid1** [comment]
this should be covered (and also be invalidated) by https://github.com/ceph/ceph/pull/34061 I believe @sebastian-philipp ?

**Comment by sebastian-philipp** [comment]
> this should be covered (and also be invalidated) by #34061 I believe @sebastian-philipp ?

your PR is independent should be merged as well!

**Comment by smithfarm** [comment]
I have the following service_spec:

```
    master: ++ cat service_spec_osd.yml
    master: service_type: osd
    master: placement:
    master:     hosts:
    master:         - master
    master: service_id: generic_osd_deployment
    master: data_devices:
    master:     all: True
```

but it doesn't work:

```
    master: ++ ceph orch apply osd -i service_spec_osd.yml
    master: Error EINVAL: Traceback (most recent call last):
<BLA, BLA>
    master: TypeError: type object argument after ** must be a mapping, not str
```

Will this PR fix it?

**Comment by jschmid1** [comment]
> Will this PR fix it?

yep

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (jschmid1) [reply src/python-common/ceph/deployment/service_spec.py]: comming back to this, I don't see the necessity to do that anymore. Do you?

Comment (sebastian-philipp) [reply src/python-common/ceph/deployment/service_spec.py]: As long as `HostPlacementSpec.from_json(HostPlacementSpec(...).to_json())` works, I'm happy!

Comment (jschmid1) [reply src/python-common/ceph/deployment/service_spec.py]: covered in the next batch of tests

**Comment by jschmid1** [comment]
oh really..

```
Traceback:
ceph/tests/test_service_spec.py:3: in <module>
    import yaml
E   ModuleNotFoundError: No module named 'yaml'
```

**Comment by sebastian-philipp** [comment]
needs rebase, sorry!

**Comment by jschmid1** [comment]
> needs rebase, sorry!

done

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-04-02_08:44:54-rados-wip-swagner-testing-2020-04-01-1304-distro-basic-smithi/

**Comment by smithfarm** [comment]
I could not find any octopus backport. Will it be staged?

**Comment by smithfarm** [comment]
I tested this with:

```
    master: ++ cat service_spec_osd.yml
    master: service_type: osd
    master: placement:
    master:     hosts:
    master:         - 'node1*'
    master:         - 'node2*'
    master:         - 'node3*'
    master: service_id: generic_osd_deployment
    master: data_devices:
    master:     all: True
```

and got:

```
    master: ++ ceph orch apply osd -i service_spec_osd.yml
    master: Error ENOENT: name component must include only a-z, 0-9, and -
```

Glob-matching of hosts is no longer supported?

**Comment by sebastian-philipp** [comment]
```yaml
service_type: osd
service_id: generic_osd_deployment
placement:
  host_pattern: 'node[1-3]*'
data_devices:
  all: True
```

See https://docs.ceph.com/docs/master/mgr/orchestrator/#placement-by-pattern-matching

**Comment by sebastian-philipp** [comment]
> I could not find any octopus backport. Will it be staged?

well spotted!

**Pull request merged**